components/org.wso2.carbon.identity.application.authenticator.push/src/main/java/org/wso2/carbon/identity/application/authenticator/push/PushAuthenticator.java

```diff
@@ -0,0 +1,481 @@
+ * under the License.
+ *
+ */
+
+package org.wso2.carbon.identity.application.authenticator.push;
+
+import org.apache.commons.lang.StringUtils;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.json.simple.JSONArray;
+import org.json.simple.JSONObject;
+import org.wso2.carbon.identity.application.authentication.framework.AbstractApplicationAuthenticator;
+import org.wso2.carbon.identity.application.authentication.framework.FederatedApplicationAuthenticator;
+import org.wso2.carbon.identity.application.authentication.framework.context.AuthenticationContext;
+import org.wso2.carbon.identity.application.authentication.framework.exception.AuthenticationFailedException;
+import org.wso2.carbon.identity.application.authentication.framework.inbound.InboundConstants;
+import org.wso2.carbon.identity.application.authentication.framework.model.AuthenticatedUser;
+import org.wso2.carbon.identity.application.authenticator.push.cache.AuthContextCache;
+import org.wso2.carbon.identity.application.authenticator.push.cache.AuthContextCacheEntry;
+import org.wso2.carbon.identity.application.authenticator.push.cache.AuthContextcacheKey;
+import org.wso2.carbon.identity.application.authenticator.push.device.handler.DeviceHandler;
+import org.wso2.carbon.identity.application.authenticator.push.device.handler.exception.PushDeviceHandlerClientException;
+import org.wso2.carbon.identity.application.authenticator.push.device.handler.exception.PushDeviceHandlerServerException;
+import org.wso2.carbon.identity.application.authenticator.push.device.handler.impl.DeviceHandlerImpl;
+import org.wso2.carbon.identity.application.authenticator.push.device.handler.model.Device;
+import org.wso2.carbon.identity.application.authenticator.push.dto.AuthDataDTO;
+import org.wso2.carbon.identity.application.authenticator.push.dto.impl.AuthDataDTOImpl;
+import org.wso2.carbon.identity.application.authenticator.push.internal.PushAuthenticatorServiceComponent;
+import org.wso2.carbon.identity.application.authenticator.push.notification.handler.impl.FirebasePushNotificationSenderImpl;
+import org.wso2.carbon.identity.application.authenticator.push.validator.PushJWTValidator;
+import org.wso2.carbon.identity.application.common.model.Property;
+import org.wso2.carbon.identity.core.util.IdentityTenantUtil;
+import org.wso2.carbon.identity.core.util.IdentityUtil;
+import org.wso2.carbon.user.api.UserStoreException;
+import org.wso2.carbon.user.api.UserStoreManager;
+import org.wso2.carbon.user.core.UserCoreConstants;
+import org.wso2.carbon.user.api.UserRealm;
+import org.wso2.carbon.user.core.service.RealmService;
+
+import java.io.IOException;
+import java.net.URLEncoder;
+import java.nio.charset.StandardCharsets;
+import java.sql.SQLException;
+import java.util.*;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import ua_parser.Parser;
+import ua_parser.Client;
```

Comment:
consider replacing with existing libs

Reply:
Imports for the replacement will remain the same. The group ID will be replaced in the pom.xml and will appear in a future PR.